### PR TITLE
Add institution programs

### PR DESCRIPTION
**Pull request by devin-mccurdy**
Programs need to be versioned and archived. Added `InstitutionProgram` which is a new model that is used when generating a new version.

The existing `programs` table will be used for the data from the uploaded CSV only.

Database migration PRs: 

- https://github.com/department-of-veterans-affairs/gibct-data-service/pull/442
- https://github.com/department-of-veterans-affairs/gibct-data-service/pull/447

https://app.zenhub.com/workspaces/vft-59c95ae5fda7577a9b3184f8/issues/department-of-veterans-affairs/vets.gov-team/19680

### CODE REVIEW

**Review by zurbergram — APPROVED** [review]
LGTM

**Review by delli-gatti-michael — APPROVED** [review]
Approved by QA

File: app/models/institution.rb [hunk 1]
```diff
@@ -161,7 +161,9 @@ class Institution < ActiveRecord::Base
   has_many :yellow_ribbon_programs, dependent: :destroy
   has_many :school_certifying_officials, -> { order 'priority, last_name' },
            primary_key: :facility_code, foreign_key: 'facility_code'
-  has_many :programs, -> { order(description: :asc) }, primary_key: :facility_code, foreign_key: 'facility_code'
+  has_many :institution_programs, -> { order(description: :asc) },
```
Comment (jholton) [inline app/models/institution.rb]: You also need to account for `version` in this `has_many`.  i.e. the relationship is built on a compound key of `facility_code` and `version`

File: spec/models/institution_builder_spec.rb [hunk 2]
```diff
@@ -891,6 +891,17 @@
       end
     end
 
+    describe 'when generating institution programs' do
+      let(:institution) { institutions.find_by(zip: zipcode_rate.zip_code) }
```
Comment (jholton) [inline spec/models/institution_builder_spec.rb]: I don't see this `institution` used anywhere.

File: spec/models/institution_builder_spec.rb [hunk 3]
```diff
@@ -891,6 +891,17 @@
...
+      let(:institution_program) { InstitutionProgram.first }
+
+      it 'properly generates institution programs from weams data' do
+        create :program
+        InstitutionBuilder.run(user)
+        expect(institution_program).not_to eq(nil)
```
Comment (jholton) [inline spec/models/institution_builder_spec.rb]: It's unclear to me that `InstitutionBuilder.run` caused `institution_program` to not be `nil`.
Perhaps we can more clearly test that `InstitutionBuilder.run` is causing `InstitutionProgram` to be created.
Something like:
```
     it 'properly generates institution programs from weams data' do
        create :program
        expect { InstitutionBuilder.run(user) }.to change{InstitutionProgram.count}.from(0).to(1)
      end
```

File: app/models/institution_builder.rb [hunk 4]
```diff
@@ -502,4 +503,30 @@ def self.add_extension_campus_type(version_number)
...
+
+  def self.build_institution_programs(version_number)
+    str = <<-SQL
+      INSERT INTO institution_programs (
+        facility_code,
+        program_type,
+        description,
+        full_time_undergraduate,
+        graduate,
+        full_time_modifier, length_in_hours,
```
Comment (jholton) [inline app/models/institution_builder.rb]: Nitpick: can we put these on two separate lines like the others?

File: app/models/institution_builder.rb [hunk 5]
```diff
@@ -502,4 +503,30 @@ def self.add_extension_campus_type(version_number)
...
+        version
+      )
+      SELECT
+        facility_code,
+        program_type,
+        description,
+        full_time_undergraduate,
+        graduate,
+        full_time_modifier, length,
```
Comment (jholton) [inline app/models/institution_builder.rb]: Nitpick: can we put these on two separate lines like the others?

**Review by jholton — CHANGES_REQUESTED** [review]

File: app/models/institution_builder.rb (hunk 5 above)
Comment (dmccurdybah) [reply app/models/institution_builder.rb]: Yes!  Sorry, I was just getting addicted to green code climate checkboxes

File: app/models/institution.rb (hunk 1 above)
Comment (dmccurdybah) [reply app/models/institution.rb]: Thanks!  This made me realize that I also made a mistake with the index on the migration PR (new PR coming soon...)

**Comment by devin-mccurdy** [comment]
@jholton Additional DB migrations are in https://github.com/department-of-veterans-affairs/gibct-data-service/pull/447

File: spec/models/institution_builder_spec.rb [hunk 6]
```diff
@@ -891,6 +891,15 @@
       end
     end
 
+    describe 'when generating institution programs' do
+      let(:institution_program) { InstitutionProgram.first }
+
+      it 'properly generates institution programs from weams data' do
+        create :program
+        expect { InstitutionBuilder.run(user) }.to change { InstitutionProgram.count }.from(0).to(1)
```
Comment (jholton) [inline spec/models/institution_builder_spec.rb]: Can we get one `expect` verifying the `version` is properly created?
I think it'd be:
`expect(institution_program.version).to eq(Version.current_preview.number)`

**Review by jholton — APPROVED** [review]
LGTM

**Pull request merged**